### ISSUE TITLE
Give form input elements a name for accessibility

### ISSUE THREAD

**Issue by PaulMorris** [SolutionGuidance/psm#554]
Part of issue #510 , we should make sure each form input element has a name for accessibility.  Here is the relevant HTML_CodeSniffer error:
> This text input element does not have a name available to an accessibility API. Valid names are: label element, title attribute, aria-label attribute, aria-labelledby attribute.
> Success Criterion:
> [4.1.2: Name, Role, Value](http://www.w3.org/TR/WCAG20/#ensure-compat-rsv)
> Suggested Techniques:
> [H91](http://www.w3.org/TR/WCAG20-TECHS/H91)

And also this error (they often appear together):
> This form field should be labelled in some way. Use the label element (either with a "for" attribute or 
> wrapped around the form field), or "title", "aria-label" or "aria-labelledby" attributes as appropriate.
> Success Criterion:
> [1.3.1: Info and Relationships](http://www.w3.org/TR/WCAG20/#content-structure-separation-programmatic)
> Suggested Techniques:
> [F68](http://www.w3.org/TR/WCAG20-TECHS/F68)

**Comment by PaulMorris** [comment]
Discussed options for automating this with @cecilia-donnelly and we came up with this approach:
1. Look for a <label> element and get the text inside it.  (Cleaning it up as needed, e.g. if it has any tags like <em> or <span> in it)
2. Look for the next <input> element
3. Check to see if it has a title attribute.  If not, add one using the text from the label.

There will be edge cases with radio buttons, etc. but still worth it to write a script to do the above.

**Comment by cecilia-donnelly** [comment]
In addition to `<input>` elements, we need to handle `<select>` elements.  @PaulMorris, I expect each `<option>` in a `<select>` needs its own `name` or `title`?

**Comment by PaulMorris** [comment]
@cecilia-donnelly Ah, yes, we do need to add title attributes to `<select>` elements and not just `<input>`.  Fortunately, the `<option>` elements in a `<select>` do not need to be modified.  (The `<option>` elements have never shown up as having this error for me.)

**Comment by slifty** [comment]
Hey folks -- I'm a little concerned with the duplication of label values from a DRY perspective; I know from @cecilia-donnelly's note to me that in many cases where a label exists the input doesn't have an id attribute.  Would it be an appropriate solution to add an id attribute in these cases so we can properly link the labels using a for=""?

There are a few reasons I think this would be a cleaner solution once finished:

1. Screen readers would know the two elements are linked, which I imagine could prevent issues where the screen reader reads both the label AND the title
2. Semantic association between associated elements is always nice (this is just a more general version of 1)
3. If the label needs to be changed, you don't have to change it twice / we remove the risk of having the screen reader version rot when the visible version is changed.

For inputs / selects without labels we will need to use title, which we may want to eventually improve by adding labels in future.

**Comment by PaulMorris** [comment]
Hey @slifty,  That makes sense to me.  You make a good case.  It would require creating unique id attributes for elements that don't have them yet, avoiding duplication of existing ids already in use by CSS/JS.  So that approach entails a more complex implementation process, but with a better result, DRY-wise.

**Comment by PaulMorris** [comment]
This was addressed in PR #583.  Closing.
